net/host.go

```diff
@@ -308,86 +334,113 @@ func (h *host) addrInfo() peer.AddrInfo {
 	}
 }
 
-func (h *host) writeToStream(s libp2pnetwork.Stream, msg Message) error {
+func writeStreamMessage(s io.Writer, msg Message, peerID peer.ID) error {
 	encMsg, err := msg.Encode()
 	if err != nil {
 		return err
 	}
 
-	msgLen := uint64(len(encMsg))
-	lenBytes := uint64ToLEB128(msgLen)
-	encMsg = append(lenBytes, encMsg...)
+	err = binary.Write(s, binary.LittleEndian, uint32(len(encMsg)))
+	if err != nil {
+		return err
+	}
 
 	_, err = s.Write(encMsg)
 	if err != nil {
 		return err
 	}
 
-	log.Debug(
-		"Sent message to peer=", s.Conn().RemotePeer(), " type=", msg.Type(),
-	)
+	log.Debugf("Sent message to peer=%s type=%s", peerID, msg.Type())
 
 	return nil
 }
 
-// readStream reads from the stream into the given buffer, returning the number of bytes read
-func readStream(stream libp2pnetwork.Stream, buf []byte) (int, error) {
-	if stream == nil {
-		return 0, errNilStream
+// readStreamMessage reads the 4-byte LE size header and message body returning the
+// message body bytes. io.EOF is returned if the stream is closed before any bytes
+// are received. If a partial message is received before the stream closes,
+// io.ErrUnexpectedEOF is returned.
+func readStreamMessage(s io.Reader) (Message, error) {
+	if s == nil {
+		return nil, errNilStream
 	}
 
-	var (
-		tot int
-	)
-
-	length, bytesRead, err := readLEB128ToUint64(stream, buf[:1])
+	lenBuf := make([]byte, 4) // uint32 size
+	n, err := io.ReadFull(s, lenBuf)
 	if err != nil {
-		return bytesRead, fmt.Errorf("failed to read length: %w", err)
-	}
-
-	if length == 0 {
-		return 0, nil
+		if isEOF(err) {
+			if n > 0 {
+				err = io.ErrUnexpectedEOF
+			} else {
+				err = io.EOF
+			}
+		}
+		return nil, err
 	}
+	msgLen := binary.LittleEndian.Uint32(lenBuf)
 
-	if length > uint64(len(buf)) {
-		log.Warnf("received message with size greater than allocated message buffer: msg size=%d, buffer size=%d",
-			length, len(buf))
-		return 0, fmt.Errorf("message size greater than allocated message buffer: got %d", length)
+	if msgLen > maxMessageSize {
+		log.Warnf("Received message longer than max allowed size: msg size=%d, max=%d",
+			msgLen, maxMessageSize)
+		return nil, fmt.Errorf("message size %d too large", msgLen)
 	}
 
-	tot = 0
-	for i := 0; i < maxReads; i++ {
-		n, err := stream.Read(buf[tot:])
-		if err != nil {
-			return n + tot, err
-		}
-
-		tot += n
-		if tot == int(length) {
-			break
+	msgBuf := make([]byte, msgLen)
+	_, err = io.ReadFull(s, msgBuf)
+	if err != nil {
+		if isEOF(err) {
+			err = io.ErrUnexpectedEOF
```

Comment:
what's the point of masking the underlying error here? why not wrap the error?

Reply:
There isn't any additional context in the underlying varieties of EOF that we care about. We want to treat them as the same regardless of what type of reader or stream is being used. The important distinction that we care about is: Was the connection closed between messages (frequently not an error), or was it closed in the middle of a message (almost always an error).